src/ipips/ipip-0431.md

```diff
@@ -0,0 +1,145 @@
+---
+title: "IPIP-0431: Opt-in Extensible CAR Metadata on Trustless Gateway"
+date: 2023-08-08
+ipip: proposal
+editors:
+  - name: Miroslav Bajtoš
+    github: bajtos
+    affiliation:
+        name: Protocol Labs
+        url: https://protocol.ai/
+relatedIssues:
+  - https://github.com/filecoin-project/boost/issues/1597
+order: 431
+tags: ['ipips']
+---
+
+## Summary
+
+Define an optional enhancement of the CARv1 stream that allows a Gateway server to provide
+additional metadata about the CARv1 response. Introduce a new content type that allows the client
+and the server to signal or negotiate the inclusion of extra metadata.
+
+## Motivation
+
+SPARK is a Filecoin Station module that measures the reputation of Storage Providers by periodically
+retrieving a random CID. Since both SPs and SPARK nodes are permissionless, and Proof of Retrieval
+is an unsolved problem, we need a way to verify that a SPARK node retrieved the given CID from the
+given SP. To enable that, we need the Trustless Gateway serving the retrieval request to include a
+retrieval attestation after the entire response was sent to the client.
+
+Aside from this specific use case, the IPFS Ecosystem at large has no reliable
+mechanism to signal that a CAR file transmission over HTTP completed successfully.
```

Comment:
Can you elaborate?

We already have a multicodec for CAR. Couldn't you retrieve a CAR file from a gateway by CID e.g. https://w3s.link/ipfs/bagbaierabxhdw7wglmlehzgobjuoq3v3bdv64iagjdhu74ysjvdecrezxldq - you don't need to signal successful transmission if the content hashes to the same CID.

If you're using the graph API (`?format=car` or `Accept: application/vnd.ipld.car`) you're being specific about what you want in the request and the client is verifying the blocks...and so they know if the transmission over HTTP completed successfully or not...right?

Reply:
I think there are a couple rough edges that motivated the desire to have additional signaling / metadata here:

1. In a Filecoin deal or car uploaded that has links to blocks that don't exist. How do you differentiate if a requested car has all the blocks that it should have, or if a missing link in the traversal was skipped / incorrectly transmitted?
2. If the client doesn't have all the codecs of all blocks, so can't parse links in blocks / "follow" the traversal or structure of the car it has asked for, how does it know it got the right blocks?

Reply:
> We already have a multicodec for CAR. Couldn't you retrieve a CAR file from a gateway by CID

Where do we get the "CAR CID" from? 
It does not exist as a concept anywhere in the specs related to retrieval or routing.

AFAIK in the majority of real world use cases:
- client does not know the "CAR CID" because CARv1 is not deterministic ([example](https://specs.ipfs.tech/http-gateways/trustless-gateway/#car-roots), [example](https://specs.ipfs.tech/http-gateways/trustless-gateway/#car-format-parameters-and-determinism)), it only knows he CID of actual data
- server does not know "CAR CID" because it is generating CAR stream in streaming fashion

> the client is verifying the blocks...and so they know if the transmission over HTTP completed successfully or not...right?

This works only if there is no HTTP midleware in-between client and server, which is never the case. There is always some HTTP middleware or CDN in production.

Once you are limited to HTTP semantics, you will cache truncated responses, and the client has to be smart enough to (1) detect that (2) be able to retry in a way that does not hit the same cache.

This is why it does not work in places like Rhea/Saturn, where HTTP responses are (last time i checked) cached blindly based only on HTTP semantics without understanding internal Block/DAG structure.